File: Easy/234. Palindrome Linked List/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include<algorithm>
#include <numeric> 


using namespace std;


//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
public:
    bool isPalindrome(ListNode* head) {
        std::vector<int> num_list;

        while (head)
        {
            num_list.push_back(head->val);
            head = head->next;
        }

        int l_idx = 0;
        int r_idx = num_list.size() - 1;
        bool is_pal = true;
        while (l_idx < r_idx)
        {
            if (num_list[l_idx] == num_list[r_idx])
            {
                l_idx++;
                r_idx--;
            }
            else {
                is_pal = false;
                break;
            }

        }

        return is_pal;

    }
};



int main() {


    Solution s = Solution();

    ListNode* list_num = new ListNode();
    list_num->val = 1;

    ListNode* num1 = new ListNode(2);
    list_num->next = num1;

    ListNode* num2 = new ListNode(2);
    list_num->next->next = num2;

    ListNode* num3 = new ListNode(1);
    list_num->next->next->next = num3;


    bool missing = s.isPalindrome(list_num);


    cout << missing << endl;


    return 0;
}

File: Easy/234. Palindrome Linked List/main.py
# Definition for singly-linked list.
from typing import Optional
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        value_list=[]
        current_node=head
        while current_node.next!=None:
            value_list.append(current_node.val)
            current_node=current_node.next
        value_list.append(current_node.val)
        
        l_idx=0
        r_idx=len(value_list)-1
        is_palindrom=True
        while l_idx<r_idx:
            if value_list[l_idx]==value_list[r_idx]:
                l_idx+=1
                r_idx-=1
            else:
                is_palindrom=False
                break;
            
        return is_palindrom
            
        
        





head=ListNode(1)
t1=ListNode(2)
t2=ListNode(2)
t3=ListNode(1)
head.next=t1
head.next.next=t2
head.next.next.next=t3


s=Solution()
s.isPalindrome(head)
print(s.isPalindrome(head))

